src/main/scala/calc/Compiler.scala

```diff
@@ -61,21 +65,93 @@ object Compiler {
     ir.Hashers.hashClassDef(classDef)
   }
 
+
+
+  def findCapture(tree: Tree): Set[String] = {
+    tree match {
+      case _:Literal => Set()
+      case BinaryOp(_, l, r) => findCapture(l) ++ findCapture(r)
+      case Let(i, v, body) => findCapture(body) ++ findCapture(v) - i.name
+      case Ident(n) => Set(n)
+      case If(cond, thenp, elsep) => findCapture(cond) ++ findCapture(thenp) ++ findCapture(elsep)
+      case Closure(idents, body) => findCapture(body) -- idents.flatMap((x:Tree) => findCapture(x)).toSet
+      case Call(f, args) => findCapture(f) ++ args.flatMap((x:Tree) => findCapture(x)).toSet
+    }
+  }
+
+  var seed = 1
+  def generateName():String = {
+    seed += 1
+    "generated_"+seed.toString
+  }
+
+  def compileExpr(tree: Tree):irt.Tree ={
+    val treeType = typeInfer(tree)
+    compileExpr(tree, Map())
+  }
+
+
   /** Compile an expression tree into an IR `Tree`, which is an expression
    *  that evaluates to the result of the tree.
    *
    *  This is the main method you have to implement.
    */
-  def compileExpr(tree: Tree): irt.Tree = {
-    // TODO
+  def compileExpr(tree: Tree, typeEnv:Map[String, irtpe.Type]): irt.Tree = {
     implicit val pos = tree.pos
-
     tree match {
       case Literal(value) =>
         irt.DoubleLiteral(value)
 
-      case _ =>
-        throw new Exception(
+      case BinaryOp(op, lhs, rhs) =>
+        val jsop = op match {
+          case "+" => irt.BinaryOp.Double_+
+          case "-" => irt.BinaryOp.Double_-
+          case "*" => irt.BinaryOp.Double_*
+          case "/" => irt.BinaryOp.Double_/
+        }
+        irt.BinaryOp(jsop, compileExpr(lhs, typeEnv), compileExpr(rhs, typeEnv))
+
+      case Let(ident, value, body) =>
+        val jsV = compileExpr(value, typeEnv)
+        val jsIdent = irt.Ident(ident.name)
+        val jsId = irt.VarDef(jsIdent, jsV.tpe, mutable = false, jsV)
+        irt.Block(List(jsId, compileExpr(body, typeEnv + (ident.name -> jsV.tpe))))
+
+      case Ident(n) =>
+        try {
+          irt.VarRef(irt.Ident(n))(typeEnv(n))
+        }catch{
+          case e:Exception => throw new Exception("No such identitier!")
+        }
+
+      case If(cond, thenp, elsep) =>
+        val jsThenp = compileExpr(thenp, typeEnv)
+        val jsElsep = compileExpr(elsep, typeEnv)
+        val jsCond = compileExpr(cond, typeEnv)
+        val finalCond = irt.BinaryOp(irt.BinaryOp.!==, irt.Unbox(jsCond,'D'), irt.DoubleLiteral(0.0))
```

Comment:
There should be no need for `Unbox` here, as the type checker asserted that `jsCond` is a double.